File: ytrue-infra/ytrue-infra-core/src/main/java/com/ytrue/infra/core/util/AssertUtil.java
package com.ytrue.infra.core.util;

import cn.hutool.core.collection.CollUtil;
import cn.hutool.core.util.NumberUtil;
import cn.hutool.core.util.ObjectUtil;
import cn.hutool.core.util.StrUtil;
import com.ytrue.infra.core.base.IServerResponseInfo;
import com.ytrue.infra.core.excptions.AssertInvalidArgumentException;

import java.util.Collection;
import java.util.Objects;

/**
 * @author ytrue
 * @date 2022/4/21 16:44
 * @description 断言工具
 */
public class AssertUtil {

    public static void notNull(Object object, IServerResponseInfo responseCode) {
        if (Objects.isNull(object)) {
            reportInvalidArgument(responseCode);
        }
    }

    public static void isNull(Object object, IServerResponseInfo responseCode) {
        if (Objects.nonNull(object)) {
            reportInvalidArgument(responseCode);
        }
    }

    public static void objectEquals(Object object1, Object object2, IServerResponseInfo responseCode) {
        if (ObjectUtil.notEqual(object1, object2)) {
            reportInvalidArgument(responseCode);
        }
    }

    public static void objectNotEquals(Object object1, Object object2, IServerResponseInfo responseCode) {
        if (ObjectUtil.equal(object1, object2)) {
            reportInvalidArgument(responseCode);
        }
    }


    public static void numberEquals(Number number1, Number number2, IServerResponseInfo responseCode) {
        if (!NumberUtil.equals(number1, number2)) {
            reportInvalidArgument(responseCode);
        }
    }

    public static void numberNotEquals(Number number1, Number number2, IServerResponseInfo responseCode) {
        if (NumberUtil.equals(number1, number2)) {
            reportInvalidArgument(responseCode);
        }
    }


    public static <T extends Number & Comparable<? super T>> void lessThanEq(T value, T limit, IServerResponseInfo responseCode) {
        if (value.compareTo(limit) > 0) {
            reportInvalidArgument(responseCode);
        }
    }


    public static <T extends Number & Comparable<? super T>> void lessThan(T value, T limit, IServerResponseInfo responseCode) {
        if (value.compareTo(limit) >= 0) {
            reportInvalidArgument(responseCode);
        }
    }

    public static <T extends Number & Comparable<? super T>> void greaterThanEq(T value, T limit, IServerResponseInfo responseCode) {
        if (value.compareTo(limit) < 0) {
            reportInvalidArgument(responseCode);
        }
    }

    public static <T extends Number & Comparable<? super T>> void greaterThan(T value, T limit, IServerResponseInfo responseCode) {
        if (value.compareTo(limit) <= 0) {
            reportInvalidArgument(responseCode);
        }
    }


    public static void isTrue(boolean expression, IServerResponseInfo responseCode) {
        if (!expression) {
            reportInvalidArgument(responseCode);
        }
    }

    public static void isFalse(boolean expression, IServerResponseInfo responseCode) {
        if (expression) {
            reportInvalidArgument(responseCode);
        }
    }


    public static void strIsEmpty(CharSequence value, IServerResponseInfo responseCode) {
        if (StrUtil.isNotEmpty(value)) {
            reportInvalidArgument(responseCode);
        }
    }


    public static void strIsNotEmpty(CharSequence value, IServerResponseInfo responseCode) {
        if (StrUtil.isEmpty(value)) {
            reportInvalidArgument(responseCode);
        }
    }

    public static void collectionIsNotEmpty(Collection<?> collection, IServerResponseInfo responseCode) {
        if (CollUtil.isEmpty(collection)) {
            reportInvalidArgument(responseCode);
        }
    }


    public static void collectionIsEmpty(Collection<?> collection, IServerResponseInfo responseCode) {
        if (CollUtil.isNotEmpty(collection)) {
            reportInvalidArgument(responseCode);
        }
    }

    private static void reportInvalidArgument(IServerResponseInfo responseCode) {
        throw new AssertInvalidArgumentException(responseCode);
    }
}

File: ytrue-admin-server-csharp/Ytrue.Infrastructures/Utilities/AssertUtil.cs
using Ytrue.Infrastructures.Bases;
using Ytrue.Infrastructures.Response;

namespace Ytrue.Infrastructures.Utilities;

public class AssertUtil
{
    public static void NotNull(object? o, IServerResponseInfo responseCode)
    {
        if (o == null)
        {
            ReportInvalidArgument(responseCode);
        }
    }

    public static void IsNull(object? o, IServerResponseInfo responseCode)
    {
        if (o != null)
        {
            ReportInvalidArgument(responseCode);
        }
    }


    public static void IsTrue(bool b, IServerResponseInfo responseCode)
    {
        if (!b)
        {
            ReportInvalidArgument(responseCode);
        }
    }


    public static void IsFalse(bool b, IServerResponseInfo responseCode)
    {
        if (b)
        {
            ReportInvalidArgument(responseCode);
        }
    }


    public static void CollectionIsNotEmpty<T>(IEnumerable<T>? collection, IServerResponseInfo responseCode)
    {
        // 如果集合为 null 或者不包含任何元素
        if (collection == null || !collection.Any())
        {
            ReportInvalidArgument(responseCode);
        }
    }


    public static void CollectionIsEmpty<T>(IEnumerable<T>? collection, IServerResponseInfo responseCode)
    {
        // 如果集合为 null 或者不包含任何元素
        if (!(collection != null && collection.Any()))
        {
            ReportInvalidArgument(responseCode);
        }
    }


    private static void ReportInvalidArgument(IServerResponseInfo responseCode)
    {
        throw new AssertInvalidArgumentException(responseCode);
    }

    private class AssertInvalidArgumentException : BaseCodeException
    {
        public AssertInvalidArgumentException(IServerResponseInfo responseCode) : base(responseCode)
        {
        }

        public AssertInvalidArgumentException(string? message) : base(message)
        {
        }
    }
}